### PR TITLE
Initial version: SQLite UDF API

### PR DESCRIPTION
**Pull request by tiagokepe**
This PR is a proof of "concept" for the support of SQLite UDF API #1792.

For now, it is just working with the _int32_ type, but when all relevant points are defined (e.g., casting back and forth the _sqlite_values_), other types can be support straightforward.

Refinements and much work are still needed, and suggestions/improvements are welcome.

### CODE REVIEW

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+struct sqlite3_value {
+    union MemValue {
+        double r;           /* Real value used when MEM_Real is set in flags */
+        int64_t i;          /* Integer value used when MEM_Int is set in flags */
+        int nZero;          /* Extra zero bytes when MEM_Zero and MEM_Blob set */
+        const char *zPType; /* Pointer type when MEM_Term|MEM_Subtype|MEM_Null */
+    } u;
+    // uint16_t flags;
+    PhysicalType type;
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:15]: Rather than using the physical type here, we can use a different enum. Note that the types that need to be stored in here are rather limited. If we look at the [sqlite3_value_type function](http://sqlite.org/c3ref/value_blob.html), we see the following types:

```c
SQLITE_INTEGER, SQLITE_FLOAT, SQLITE_TEXT, SQLITE_BLOB, or SQLITE_NULL.
```

This corresponds with the methods used for extracting the actual data value from the sqlite3_value wrapper:

```c
const void *sqlite3_value_blob(sqlite3_value*);
double sqlite3_value_double(sqlite3_value*);
int sqlite3_value_int(sqlite3_value*);
sqlite3_int64 sqlite3_value_int64(sqlite3_value*);
void *sqlite3_value_pointer(sqlite3_value*, const char*);
const unsigned char *sqlite3_value_text(sqlite3_value*);
```

In our sqlite3_value, I would store the same types using the same type enum (`int type`).

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+struct sqlite3_context {
+    sqlite3_value result;
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:19]: The list of functions that can operate on the sqlite3_context is [here](https://sqlite.org/c3ref/context.html). We will need to implement all of these, although some of them can be stubs for now (e.g. `sqlite3_aggregate_context` is a stub).

Of interest for the contents of the `sqlite3_context` are:
* [sqlite3_set/getauxdata](https://sqlite.org/c3ref/get_auxdata.html)
* [sqlite3_result_error](https://sqlite.org/c3ref/result_blob.html)
* [sqlite3_user_data](https://sqlite.org/c3ref/user_data.html)
* [sqlite3_context_db_handle](https://sqlite.org/c3ref/context_db_handle.html)

All of the necessary information to implement these functions should be stored within the context.

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		sqlite3_context context;
+		int argc = 1;
+		for(idx_t i=0; i < vec_sqlite.size(); ++i) {
+			sqlite3_value *value = &vec_sqlite[i];
+
+			//calling the SQLite3 UDF
+			sqlite_udf(&context, argc, &value);
+
+			//casting the result sqlite value to string_t
+			if(TypeIsInteger(context.result.type)) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:33]: It would be faster to do the result conversion in a vectorized manner as well, i.e. store all of the resulting values in an array and then bulk convert them in a loop.

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+scalar_function_t SQLiteUDFWrapper::CreateBinarySQLiteFunction(scalar_sqlite_udf_t sqlite_udf) {
+	scalar_function_t udf_function = [=](DataChunk &args, ExpressionState &state, Vector &result) -> void {
+		assert(args.ColumnCount() == 1);
+
+		VectorData vec_data;
+		args.data[0].Orrify(args.size(), vec_data);
+
+		vector<sqlite3_value> vec_sqlite(args.size());
+		CastSQLite::ToVector(args.data[0].GetType(), vec_data, vec_sqlite);
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:15]: I think it should be possible to create a single function that handles an arbitrary number of arguments, rather than needing to create different functions for 1 argument, 2 arguments, 3 arguments, etc as we did in the vectorized UDFs.

The reason we did this in the vectorized UDFs is because the template specialisation on the number of arguments allows the function call to be inlined. This is significantly faster, because we save the function call per value.

For the SQLite UDFs, we unfortunately have no choice. The format of the function (a function pointer) does not allow us to inline it, and hence specialising on the number of arguments does not get us the same performance boost. 

Instead, you could create a single function that handles all the argument types here, which would perform a loop over the input vectors (`args.ColumnCount()` and constructs the `sqlite3_value` arguments for each of the function invocations.

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 5]
```diff
@@ -1120,8 +1123,17 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
                             void (*xFunc)(sqlite3_context *, int, sqlite3_value **),
                             void (*xStep)(sqlite3_context *, int, sqlite3_value **),
                             void (*xFinal)(sqlite3_context *)) {
-	// fprintf(stderr, "sqlite3_create_function: unsupported.\n");
-	return -1;
+
+	string fname = string(zFunctionName);
+
+	//Unary function for now
+	if(nArg == 1 && xFunc) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1130]: See above, I think the specialisation on `nArg` can be avoided.

**Review by Mytherin — COMMENTED** [review]
Thanks for the work! This is definitely on the right track to what we are looking for. Some comments:

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 6]
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
...
-const void *sqlite3_value_blob(sqlite3_value *) {
-	return nullptr;
+const void *sqlite3_value_blob(sqlite3_value *pVal) {
+	return sqlite3_value_text(pVal);
 }
 
-double sqlite3_value_double(sqlite3_value *) {
-	return 0;
+double sqlite3_value_double(sqlite3_value *pVal) {
+	if (pVal->type == SQLiteTypeValue::FLOAT) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1437]: I believe we should try to cast the value in the case the underlying value is an integer or text value, see e.g. [the docs](https://sqlite.org/c3ref/value_blob.html):

> As long as the input parameter is correct, these routines can only fail if an out-of-memory error occurs during a format conversion. Only the following subset of interfaces are subject to out-of-memory errors:
sqlite3_value_blob()
sqlite3_value_text()
sqlite3_value_text16()
sqlite3_value_text16le()
sqlite3_value_text16be()
sqlite3_value_bytes()
sqlite3_value_bytes16()

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 7]
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
...
 
-int sqlite3_value_int(sqlite3_value *) {
-	return 0;
+int sqlite3_value_int(sqlite3_value *pVal) {
+	return sqlite3_value_int64(pVal);
 }
 
-sqlite3_int64 sqlite3_value_int64(sqlite3_value *) {
+sqlite3_int64 sqlite3_value_int64(sqlite3_value *pVal) {
+	if (pVal->type == SQLiteTypeValue::INTEGER) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1448]: Same here, I believe we should add a cast here if the underlying value type is different.

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 8]
```diff
@@ -1386,27 +1429,38 @@ SQLITE_API int sqlite3_result_zeroblob64(sqlite3_context *, sqlite3_uint64 n) {
...
 void *sqlite3_value_pointer(sqlite3_value *, const char *) {
 	return nullptr;
 }
 
-const unsigned char *sqlite3_value_text(sqlite3_value *) {
+const unsigned char *sqlite3_value_text(sqlite3_value *pVal) {
+	if (pVal && (pVal->type == SQLiteTypeValue::TEXT || pVal->type == SQLiteTypeValue::BLOB)) {
+		// We don't need to append \0 at the end of string
+		// To know the number of chars or bytes, call function sqlite3_value_bytes()
+		return (const unsigned char *)pVal->str_t.GetDataUnsafe();
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1462]: Same here.

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/include/sqlite3_value_type.hpp [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+enum class SQLiteTypeValue : uint8_t { INTEGER = 1, FLOAT = 2, TEXT = 3, BLOB = 4, NULL_VALUE = 5 };
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/include/sqlite3_value_type.hpp:3]: Could you wrap this in the duckdb namespace?

File: tools/sqlite3_api_wrapper/test/include/sqlite_stmt_wrapper.hpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+		Finalize();
+		return sqlite3_prepare_v2(db, zSql, nByte, &stmt, pzTail);
+	}
+
+	/**
+	 * Execute a prepated statment previously "groomed" and print the result
+	 */
+	int ExecutePreparedStmt() {
+		if (!stmt) {
+			fprintf(stderr, "There is no prepated statement: Prepare(...) must be invoked firstly.\n");
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/test/include/sqlite_stmt_wrapper.hpp:25]: prepared

File: tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	REQUIRE(db_w.Execute("CREATE TABLE integers(i INTEGER)"));
+	for (int i = -5; i <= 5; ++i) {
+		// Insert values: -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5
+		REQUIRE(db_w.Execute("INSERT INTO integers VALUES (" + std::to_string(i) + ")"));
+	}
+
+	// create sqlite udf
+	REQUIRE(sqlite3_create_function(db_w.db, "multiply10", 1, 0, nullptr, &multiply10, nullptr, nullptr) == SQLITE_OK);
+	REQUIRE(db_w.Execute("SELECT multiply10(i) FROM integers"));
+	REQUIRE(db_w.CheckColumn(0, {"-50", "-40", "-30", "-20", "-10", "0", "10", "20", "30", "40", "50"}));
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp:31]: Could you add a test with NULL inputs?

File: tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+}
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/test/test_sqlite3_udf_api_wrapper.cpp:32]: Could you add a test that results in an error (e.g. a division operator, that throws an error on zero on the rhs)?

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 13]
```diff
@@ -1323,16 +1343,26 @@ SQLITE_API char *sqlite3_win32_unicode_to_utf8(LPCWSTR zWideText) {
 #endif
 
 // TODO complain
-SQLITE_API void sqlite3_result_blob(sqlite3_context *, const void *, int, void (*)(void *)) {
+SQLITE_API void sqlite3_result_blob(sqlite3_context *context, const void *blob, int n_bytes, void (*)(void *)) {
+	if (blob) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1347]: We prefer to [return early](https://medium.com/swlh/return-early-pattern-3d18a41bba8):
```cpp
if (!blob) {
   return;
}
```

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 14]
```diff
@@ -1120,8 +1125,22 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
                             void (*xFunc)(sqlite3_context *, int, sqlite3_value **),
                             void (*xStep)(sqlite3_context *, int, sqlite3_value **),
                             void (*xFinal)(sqlite3_context *)) {
-	// fprintf(stderr, "sqlite3_create_function: unsupported.\n");
-	return -1;
+
+	string fname = string(zFunctionName);
+
+	// Scalar function
+	if (xFunc) {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1132]: Same here: return early.

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 15]
```diff
@@ -1120,8 +1125,22 @@ int sqlite3_create_function(sqlite3 *db, const char *zFunctionName, int nArg, in
                             void (*xFunc)(sqlite3_context *, int, sqlite3_value **),
                             void (*xStep)(sqlite3_context *, int, sqlite3_value **),
                             void (*xFinal)(sqlite3_context *)) {
-	// fprintf(stderr, "sqlite3_create_function: unsupported.\n");
-	return -1;
+
+	string fname = string(zFunctionName);
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1129]: What if `zFunctionName` is `nullptr`?

File: tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp [hunk 16]
```diff
@@ -1347,16 +1377,29 @@ SQLITE_API void sqlite3_result_error_nomem(sqlite3_context *) {
...
+SQLITE_API void sqlite3_result_null(sqlite3_context *context) {
+	context->result.type = SQLiteTypeValue::NULL_VALUE;
 }
 
-SQLITE_API void sqlite3_result_text(sqlite3_context *, const char *, int, void (*)(void *)) {
+SQLITE_API void sqlite3_result_text(sqlite3_context *context, const char *str_c, int n_chars, void (*)(void *)) {
+	if (str_c) {
+		auto utf_type = Utf8Proc::Analyze(str_c, n_chars);
+		if (utf_type == UnicodeType::INVALID) {
+			throw Exception("Text value is not valid UTF8");
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:1396]: You should not throw exceptions from within a C API function. C programs cannot deal with exceptions as they are a C++ function, so this will always result in unpredictable behavior and/or aborting the program.

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+		return ToVectorSQLiteValue<double>((double *)vec_data.data, size);
+	}
+	case LogicalTypeId::BLOB: {
+		return ToVectorSQLiteValueBlob((string_t *)vec_data.data, size);
+	}
+	case LogicalTypeId::VARCHAR: {
+		return ToVectorSQLiteValue<string_t>((string_t *)vec_data.data, size);
+	}
+	default:
+		throw std::runtime_error("SQLite UDF API: type is not supported!");
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:99]: Perhaps as a fallback cast the input vector to `VARCHAR`?

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+template <>
+string_t CastSQLite::GetValue(sqlite3_value input) {
+	return input.str_t;
+}
+
+//*** Cast to vectors ***********************************/
+unique_ptr<vector<sqlite3_value>> CastSQLite::ToVectorSQLite(LogicalType type, VectorData &vec_data, idx_t size) {
+	LogicalTypeId type_id = type.id();
+	switch (type_id) {
+	case LogicalTypeId::TINYINT: {
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:74]: LogicalType::NULL is missing. Could you add a test in the form of `SELECT my_sqlite_function(NULL)`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! Looking good. Some comments:

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+			bool is_constant_vector = false;
+			if (data_chunk.data[i].GetVectorType() == VectorType::CONSTANT_VECTOR) {
+				is_constant_vector = true;
+				// it's needed to change to FLAT_VECTOR for the COPY works
+				data_chunk.data[i].SetVectorType(VectorType::FLAT_VECTOR);
+			}
+
+			data_chunk.data[i].SetType(vchar_type);
+			VectorOperations::Copy(varchar_vector, data_chunk.data[i], data_chunk.size(), 0, 0);
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/cast_sqlite.cpp:41]: Instead of both casting and copying, could we create a DataChunk that either (1) performs a cast if required, or (2) references the input vector of the other data chunk if not required? That would save a copy and simplify the code here somewhat.

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "duckdb/common/types.hpp"
+#include "duckdb/common/types/string_type.hpp"
+#include "duckdb/common/types/string_heap.hpp"
+
+#include "sqlite3.h"
+#include "sqlite3_value_type.hpp"
+
+using namespace duckdb;
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/include/udf_struct_sqlite3.h:10]: Please avoid `using namespace`

File: tools/sqlite3_api_wrapper/test/include/sqlite_db_wrapper.hpp [hunk 21]
```diff
@@ -0,0 +1,66 @@
+#include "sqlite3.h"
+#include <string>
+#include <vector>
+#include "util_functions.hpp"
+
+using namespace std;
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/test/include/sqlite_db_wrapper.hpp:6]: Please avoid `using namespace`

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+		context.pFunc.pUserData = pApp; // set the function data
+
+		size_t argc = args.ColumnCount();                     // num of args for the UDF
+		vector<sqlite3_value> res_sqlite_values(args.size()); // to store the results from the UDF calls
+		vector<sqlite3_value> vec_values_to_free; // sqlite_values to free because some memory allocation has happened
+
+		// traversing the vector of sqlite values
+		for (idx_t row_idx = 0; row_idx < res_sqlite_values.size(); ++row_idx) {
+
+			sqlite3_value **argv = new sqlite3_value *[argc];
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:34]: Could we avoid allocating this for every row iteration?
Could we wrap this in a unique_ptr, to avoid having to manually call delete later on (e.g. `unique_ptr<sqlite_value[]>(new sqlite_value[argc])`)?

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+
+			// create a tuple from sqlite_values
+			for (idx_t col_idx = 0; col_idx < argc; ++col_idx) {
+				argv[col_idx] = &(*(vec_sqlite[col_idx]))[row_idx];
```
Comment (Mytherin) [inline tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:38]: Perhaps you could directly construct the sqlite3_value objects in the correct row-wise format in `CastSQLite::ToVectorsSQLiteValue`? That avoids having to move them from column-wise format to row-wise format here.

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! Looking good. Some more feedback:

File: tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp (hunk 23 above)
Comment (tiagokepe) [reply tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:38]: Hey Mark,
Thanks for the suggestions, I already implemented them all.

But I guess this one is a little problematic. I don't think so directly construct the sqlite3_value objects in the ToVectorsSQLiteValue is a better option because we still need to perform the tuple reconstruction.
- Constructing the tuple in the CastSQLite::ToVector may inhibit the vectorized cast to sqlite3_values.
- Another issue is the values of the argv (**argv) are pointers, so we still need to store sqlite values in some place, and then constructing the argv argument to point to the sqlite3_values.

P.S. I've already implemented such code, if you want I can share with you.

Comment (Mytherin) [reply tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:38]: Hey Tiago, you're correct, the fact that they are pointers complicates it somewhat and the benefits of constructing them in row-wise format in the `ToVectorsSQLiteValue` is significantly lower as we would still need to set up the pointers. I'm fine with leaving this as-is.

Could you update the PR with your new code?

Comment (tiagokepe) [reply tools/sqlite3_api_wrapper/sqlite3_udf_api/sqlite3_udf_wrapper.cpp:38]: I pushed the code and the workflow was triggered on my git fork.
But it appears that the workflow on duckdb repository was not triggered automatically.

**Comment by Mytherin** [comment]
Could you have a look at resolving the merge conflicts with master?

**Comment by Mytherin** [comment]
Thanks, looks great now!
